pokemongo_bot/cell_workers/sniper.py

```diff
@@ -0,0 +1,370 @@
+                    break
+
+        # If target exists, catch it, otherwise ignore
+        if exists:
+            self._log('Meeting {} successful!'.format(pokemon['name']))
+            self._teleport_back_and_catch(last_position, pokemon)
+        else:
+            self._error('Could not meet {}. Reasons: too far, caught, expired or fake data. Skipping...'.format(pokemon['name']))
+            self._teleport_back(last_position)
+
+        # Save target and unlock heartbeat calls
+        self._cache(pokemon)
+        self.bot.hb_locked = False
+
+        return WorkerResult.SUCCESS
+
+    def work(self):
+        targets = []
+
+        # Retrieve the targets
+        if self.mode == SnipingMode.SOCIAL:
+            targets = self._get_pokemons_from_social()
+        elif self.mode == SnipingMode.URL:
+            targets = self._get_pokemons_from_url()
+
+        # Order the targets (descending)
+        for attr in self.order_by:
+            targets.sort(key=lambda pokemon: pokemon[attr], reverse=True)
+
+        # Start sniping the first 'max_consecutive_catches' entries
+        for catch_attempt in range(len(targets)):
+            if catch_attempt < self.max_consecutive_catches:
+                self.snipe(targets[catch_attempt])
+
+                # Wait a bit if were going to snipe again
+                if self.max_consecutive_catches is not 1:
+                    time.sleep(3)
+
+        return WorkerResult.SUCCESS
+
+    def _validate_mappings(self, dictionary):
+        invalid_mappings = []
+
+        # Gather invalid mappings. If one is optional, the correct value can still be retrieved
+        for key, value in self.mappings.iteritems():
+            if key not in self.optional_mappings and value not in dictionary:
+                invalid_mappings.append(value)
+
+        # If mappings are invalid, this would lead to a wrong formatted pokemon object (blank attributes)
+        if invalid_mappings:
+            raise Exception('Invalid mapping values for {} mode: {}'.format(self.mode, ', '.join(invalid_mappings)))
+
+        # Mappings are good. No need to check again
+        self.is_mappings_good = True
+
+    def _parse_pokemons(self, pokemon_dictionary_list):
+        pokemons = []
+
+        # If were parsing for the first time, lets validate the mappings (use first item as a dictionary sample)
+        if not self.is_mappings_good and pokemon_dictionary_list:
+            # If mappings contain errors, log it instead of forwarding the exception
+            try:
+                self._validate_mappings(pokemon_dictionary_list[0])
+            except Exception as exception:
+                self._error(exception)
+                return pokemons
+
+        # Build up the pokemon. Pops are used to destroy random attribute names and keep the known ones!
+        for pokemon_dictonary in pokemon_dictionary_list:
+            pokemon_dictonary['iv'] = pokemon_dictonary.pop(self.mappings.get(ResponseMapper.IV), 0)
```

Comment:
Not sure again what you mean. This line will assign a value to the 'iv' key. This value will be the corrresponding to the IV mapping name that you have specified. Even if for some reason this key does not exist, it will return 0 and pop fill fail (do nothing). The pops are used to keep a clean dictionary with single keys. By doing that, we will avoid having something like that: { 'id': 143, 'pokemon_id': 143 ... }.